weasyprint/draw/text.py

```diff
@@ -172,15 +173,36 @@ def draw_first_line(stream, textbox, text_overflow, block_ellipsis, matrix):
             string = ''
             stream.set_font_size(font.hash, 1 if font.bitmap else font_size)
         string += '<'
-        for i in range(num_glyphs):
-            glyph_info = glyphs[i]
-            glyph = glyph_info.glyph
+        for i in range(number_of_glyphs):
+            glyph_info = glyphs_info[i]
+            glyph_id = glyph_info.glyph
             width = glyph_info.geometry.width
-            if (glyph == pango.PANGO_GLYPH_EMPTY or
-                    glyph & pango.PANGO_GLYPH_UNKNOWN_FLAG):
+
+            # Display zero-width empty glyph.
+            if glyph_id == pango.PANGO_GLYPH_EMPTY:
                 string += f'>{-width / font_size}<'
                 continue
 
+            # Display .notdef and log warning for missing glyphs.
+            if glyph_id & pango.PANGO_GLYPH_UNKNOWN_FLAG:
+                codepoint = glyph_id - pango.PANGO_GLYPH_UNKNOWN_FLAG
+                LOGGER.warning(
+                    '.notdef glyph rendered for Unicode string unsupported by fonts: '
```

Comment:
Oh I see now it’s only one code point then we don’t need to log it as a string anymore:

```py
                LOGGER.warning(
                    '.notdef glyph rendered for Unicode code point unsupported by fonts: '
                    f'U+{codepoint:04X} "{chr(codepoint)}"')
```